developer_guides/firmware/component-tutorial/tut-i-basic-fw-code.rst

```diff
@@ -0,0 +1,388 @@
+(aka devices) and their drivers, refer to :ref:`apps-component-overview`).
+
+.. code-block:: c
+
+   #include <sof/audio/component.h>
+
+   /* ...
+    */
+
+   struct comp_driver comp_amp = {
+           .type = SOF_COMP_AMP,
+           .ops = {
+                   .new = amp_new,
+                   .free = amp_free,
+                   .params = NULL,
+                   .cmd = NULL,
+                   .trigger = amp_trigger,
+                   .prepare = amp_prepare,
+                   .reset = amp_reset,
+                   .copy = amp_copy,
+                   .cache = NULL
+           },
+   };
+
+   static void sys_comp_amp_init(void)
+   {
+           comp_register(&comp_amp);
+   }
+
+   DECLARE_MODULE(sys_comp_amp_init);
+
+Note the ``type`` used for the component driver is set to the ``SOF_COMP_AMP``
+declared earlier. The API declaration is followed by a registration handler
+attached to the initialization list by ``DECLARE_MODULE()`` macro. This is all
+the infrastructure needs to know in order to find and create an instance of
+``SOF_COMP_AMP`` component.
+
+Some of the operations are left unimplemented at the moment:
+
+* ``params`` - the amplifier will do all the preparations and setup inside
+  the ``prepare`` handler and this one will not be used.
+
+* ``cmd`` - a handler to report and receive our custom run-time parameters will
+  be implemented later in :ref:`amp-run-time-params`.
+
+* ``cache`` - this handler, responsible for L1 cache operations will be
+  implemented later. It is not required by a basic example when the pipeline
+  is created on a single DSP core.
+
+Before you start implementing the handlers, add trace macros at the beginning
+of the *amp.c*. Note the ``TRACE_CLASS_AMP`` class identifier declared earlier.
+
+.. code-block:: c
+
+   #define trace_amp(__e, ...) trace_event(TRACE_CLASS_AMP, __e, ##__VA_ARGS__)
+   #define tracev_amp(__e, ...) tracev_event(TRACE_CLASS_AMP, __e, ##__VA_ARGS__)
+   #define trace_amp_error(__e, ...) \
+           trace_error(TRACE_CLASS_AMP, __e, ##__VA_ARGS__)
+
+Constructor ``amp_new()``
+=========================
+
+Add the following handler before your API declaration:
+
+.. code-block:: c
+
+   static struct comp_dev *amp_new(struct sof_ipc_comp *comp)
+   {
+           struct comp_dev *dev;
+           struct sof_ipc_comp_process *amp;
+           struct sof_ipc_comp_process *ipc_amp
+                   = (struct sof_ipc_comp_process *)comp;
+           struct amp_comp_data *cd;
+
+           dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+                         COMP_SIZE(struct sof_ipc_comp_process));
+           if (!dev)
+                   return NULL;
+
+           cd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*cd));
+           if (!cd) {
+                   rfree(dev);
+                   return NULL;
+           }
+
+           amp = (struct sof_ipc_comp_process *)&dev->comp;
```

Comment:
`amp` variable is unnecessary noise. Most components just do `memcpy_s(&dev->comp...)`

Reply:
I just verified some components, all host, volume, eq_fir do the cast and copy base ipc settings accordingly to the type of target. I actually assumed this would be more readable to follow that pattern and show what the dev->comp type is in this case.

Reply:
You are right, that's mixed - kpb, mux, selector do it directly, others as you did. However for showing what's going on it's better to leave it as you wrote.